File: python-backend/workflows/eval_execution.py
"""Workflow 3 — Eval Execution (Confluence v0.1 Script 3).

Four phases:

    1. Source data ingestion (CSV/JSON)
    2. List evaluator lists; pick or create 'support-quality'
    3. Get evaluator list by id; show contained evaluators
    4. Execute evaluation per row via client.evals.evaluate(...)
"""
from __future__ import annotations

import csv
import json
from collections.abc import AsyncIterator
from pathlib import Path
from typing import Any

from services.sdk_client import view_in_sandbox
from services.snippet_emitter import SnippetEmitter

DATA_DIR = Path(__file__).resolve().parent.parent / "data"
EVAL_LIST_NAME = "support-quality"

EVALUATOR_SPECS = [
    {"name": "action-advancement", "type": "llm_custom",
     "prompt": "Did the assistant advance the user toward their goal? 1 if yes, 0 if no."},
    {"name": "tool-selection-quality", "type": "llm_custom",
     "prompt": "Did the assistant pick the right tool? 1 if yes, 0 if no."},
    {"name": "helpfulness", "type": "llm_custom",
     "prompt": "Rate helpfulness from 1 (unhelpful) to 5 (very helpful)."},
    {"name": "latency-under-2s", "type": "code",
     "code": "def evaluate(trace):\n    return 1 if trace.get('latency_ms', 9999) < 2000 else 0\n",
     "language": "python"},
]


def _safe(fn, *args, **kwargs):
    try:
        return fn(*args, **kwargs), None
    except Exception as exc:  # noqa: BLE001
        return None, str(exc)


def _id(x):
    if isinstance(x, dict):
        return x.get("id") or "n/a"
    return getattr(x, "id", "n/a")


async def _phase_1(emitter: SnippetEmitter, client: Any, project_id: str, state: dict) -> AsyncIterator[str]:
    async for ev in emitter.phase_start("phase-1-source-data", 1, 4):
        yield ev

    async for ev in emitter.code(
        'with open("data/dataset.csv") as f:\n'
        '    rows = list(csv.DictReader(f))\n'
        'print(rows[:3])',
        log="Load eval input rows from CSV",
    ):
        yield ev

    rows: list[dict] = []
    csv_path = DATA_DIR / "dataset.csv"
    if csv_path.exists():
        with csv_path.open() as f:
            rows = list(csv.DictReader(f))
    else:
        json_path = DATA_DIR / "dataset.json"
        if json_path.exists():
            with json_path.open() as f:
                data = json.load(f)
            rows = [
                {"input": r.get("query", ""), "expected_tool": r.get("expected_tool"), "scope": "in"}
                for r in data.get("in_scope", [])
            ]

    state["rows"] = rows
    preview = json.dumps(rows[:3], ensure_ascii=False)
    async for ev in emitter.result(
        f"Loaded {len(rows)} rows; first 3: {preview}"
    ):
        yield ev

    async for ev in emitter.phase_end():
        yield ev


async def _phase_2(emitter: SnippetEmitter, client: Any, project_id: str, state: dict) -> AsyncIterator[str]:
    async for ev in emitter.phase_start("phase-2-evaluator-list", 2, 4):
        yield ev

    async for ev in emitter.code(
        "from browserstack_ai_sdk import Evaluate\n"
        "\n"
        "existing_lists = Evaluate.evaluator_list.list()",
        log="Find existing 'support-quality' list",
    ):
        yield ev
    listed, err = _safe(client.evaluate.evaluator_list.list)
    # Real shape: dict {"evaluators": [...], "totalCount": int}. Handle dict
    # AND object forms — earlier code only checked the object form, so the
    # bootstrap branch ran every time even when the list already existed.
    if isinstance(listed, dict):
        items = listed.get("evaluators") or listed.get("data") or []
    elif isinstance(listed, list):
        items = listed
    else:
        items = getattr(listed, "evaluators", None) or []
    found = next(
        (it for it in items if (it.get("name") if isinstance(it, dict) else getattr(it, "name", None)) == EVAL_LIST_NAME),
        None,
    )

    if found is None:
        async for ev in emitter.code(
            "from browserstack_ai_sdk import Evaluate\n"
            "\n"
            "# 'support-quality' missing — bootstrap inline so this workflow is\n"
            "# self-contained when Workflow 1 hasn't been run yet. Reuse evaluators\n"
            "# if they already exist in this project (server rejects duplicate names).\n"
            "existing = Evaluate.evaluator.list({\"limit\": 100})\n"
            "by_name = {e[\"name\"]: e[\"id\"] for e in (existing.get(\"evaluators\") or [])}\n"
            "evaluator_ids = []\n"
            "for spec in EVALUATOR_SPECS:\n"
            "    if spec[\"name\"] in by_name:\n"
            "        evaluator_ids.append(by_name[spec[\"name\"]])\n"
            "        continue\n"
            "    e = Evaluate.evaluator.create(spec)\n"
            "    evaluator_ids.append(e.id)\n"
            "Evaluate.evaluator_list.create({\n"
            f'    "name": "{EVAL_LIST_NAME}",\n'
            '    "evaluators": [{"evaluatorId": eid, "params": []} for eid in evaluator_ids],\n'
            "})",
            log="Bootstrap evaluators + list (reuse-if-exists)",
        ):
            yield ev
        # Reuse-if-exists: list current evaluators by name first, then only
        # create the ones that are missing. The server rejects create() with a
        # 409 if the name is already taken in this project, which would leave
        # us with an empty evaluator list and a broken eval-list create below.
        existing_listed, _ = _safe(client.evaluate.evaluator.list, {"limit": 100})
        if isinstance(existing_listed, dict):
            existing_evals = existing_listed.get("evaluators") or existing_listed.get("data") or []
        elif isinstance(existing_listed, list):
            existing_evals = existing_listed
        else:
            existing_evals = getattr(existing_listed, "evaluators", None) or []
        existing_by_name = {
            (e.get("name") if isinstance(e, dict) else getattr(e, "name", None)):
                (e.get("id") if isinstance(e, dict) else getattr(e, "id", None))
            for e in existing_evals
        }

        eval_ids: list[str] = []
        for spec in EVALUATOR_SPECS:
            if spec["name"] in existing_by_name and existing_by_name[spec["name"]]:
                eval_ids.append(existing_by_name[spec["name"]])
                continue
            if spec["type"] == "llm_custom":
                request = {
                    "type": "llm_custom",
                    "name": spec["name"],
                    "description": spec.get("description") or spec["name"],
                    "prompt": spec["prompt"],
                    "modelParams": {"provider": "openai", "model": "gpt-4o-mini", "adapter": "openai"},
                    "parameters": {
                        "score_reasoning_prompt": "Briefly explain your reasoning.",
                        "score_range_prompt": "Provide a discrete score of 0 or 1",
                    },
                }
            else:
                request = {
                    "type": "code",
                    "name": spec["name"],
                    "description": spec.get("description") or spec["name"],
                    "language": spec.get("language", "javascript"),
                    "code": spec["code"],
                }
            res, ee = _safe(client.evaluate.evaluator.create, request)
            eid = _id(res)
            if eid != "n/a":
                eval_ids.append(eid)
            else:
                async for ev in emitter.code(
                    f'# evaluator.create({spec["name"]}) failed: {ee}',
                    log=f"evaluator.create({spec['name']}) failed: {ee}",
                ):
                    yield ev
        # Real signature: evaluator_list.create(request_data) single dict;
        # each evaluator entry needs {evaluatorId, params: []}.
        created, cerr = _safe(
            client.evaluate.evaluator_list.create,
            {
                "name": EVAL_LIST_NAME,
                "evaluators": [{"evaluatorId": eid, "params": []} for eid in eval_ids],
            },
        )
        list_id = _id(created)
        state["evaluator_list_id"] = list_id
        async for ev in emitter.result(
            f"Created '{EVAL_LIST_NAME}' (id={list_id})" + (f" — {cerr}" if cerr else ""),
            view_in_sandbox={"label": "View evaluator list", "url": view_in_sandbox(project_id, "evaluator-list", list_id)},
        ):
            yield ev
    else:
        list_id = _id(found)
        state["evaluator_list_id"] = list_id
        async for ev in emitter.code(
            f'# Found existing list (id={list_id}); reusing',
            log="Reusing existing list",
        ):
            yield ev
        async for ev in emitter.result(
            f"Reused '{EVAL_LIST_NAME}' (id={list_id})",
            view_in_sandbox={"label": "View evaluator list", "url": view_in_sandbox(project_id, "evaluator-list", list_id)},
        ):
            yield ev

    async for ev in emitter.phase_end():
        yield ev


async def _phase_3(emitter: SnippetEmitter, client: Any, project_id: str, state: dict) -> AsyncIterator[str]:
    async for ev in emitter.phase_start("phase-3-evaluator-list-get", 3, 4):
        yield ev

    list_id = state.get("evaluator_list_id", "n/a")
    async for ev in emitter.code(
        "from browserstack_ai_sdk import Evaluate\n"
        "\n"
        f'fetched = Evaluate.evaluator_list.get("{list_id}")\n'
        "# Walk evaluatorConfigs[] to get each evaluatorId, then fetch full\n"
        "# evaluator rows so we can build the rich payload for Phase 4.\n"
        'configs = fetched.get("evaluatorConfigs") or []\n'
        'evaluator_ids = [c.get("evaluatorId") for c in configs if c.get("evaluatorId")]\n'
        "evaluators_for_execution = []\n"
        "for eid in evaluator_ids:\n"
        "    ev = Evaluate.evaluator.get(eid)\n"
        "    entry = {\n"
        '        "metricName": ev["name"],\n'
        '        "displayName": ev["name"],\n'
        '        "family": ev.get("family"),\n'
        '        "runtimeProvider": ev.get("runtimeProvider"),\n'
        '        "evaluatorId": eid,\n'
        '        "params": {},\n'
        "    }\n"
        '    if ev.get("code"): entry["codeText"] = ev["code"]\n'
        '    if ev.get("language"): entry["codeLanguage"] = ev["language"]\n'
        "    evaluators_for_execution.append(entry)",
        log="Walk evaluatorConfigs, fetch each evaluator, build payload for evaluation_execution",
    ):
        yield ev
    fetched, err = _safe(client.evaluate.evaluator_list.get, list_id)
    configs: list = []
    if fetched is not None:
        if isinstance(fetched, dict):
            configs = fetched.get("evaluatorConfigs") or []
        else:
            configs = getattr(fetched, "evaluatorConfigs", None) or []

    evaluator_ids: list[str] = []
    for c in configs:
        eid = c.get("evaluatorId") if isinstance(c, dict) else getattr(c, "evaluatorId", None)
        if eid:
            evaluator_ids.append(eid)

    evaluators_for_execution: list[dict] = []
    evaluator_details: list[dict] = []
    for eid in evaluator_ids:
        ev_obj, _ = _safe(client.evaluate.evaluator.get, eid)
        if ev_obj is None:
            continue
        name = ev_obj.get("name") if isinstance(ev_obj, dict) else getattr(ev_obj, "name", None)
        family = ev_obj.get("family") if isinstance(ev_obj, dict) else getattr(ev_obj, "family", None)
        runtime = ev_obj.get("runtimeProvider") if isinstance(ev_obj, dict) else getattr(ev_obj, "runtimeProvider", None)
        code_text = ev_obj.get("code") if isinstance(ev_obj, dict) else getattr(ev_obj, "code", None)
        code_lang = ev_obj.get("language") if isinstance(ev_obj, dict) else getattr(ev_obj, "language", None)
        evaluator_details.append({"id": eid, "name": name})
        entry: dict = {
            "metricName": name,
            "displayName": name,
            "family": family,
            "runtimeProvider": runtime,
            "evaluatorId": eid,
            "params": {},
        }
        if code_text:
            entry["codeText"] = code_text
        if code_lang:
            entry["codeLanguage"] = code_lang
        evaluators_for_execution.append(entry)

    state["evaluators_for_execution"] = evaluators_for_execution
    state["evaluator_details"] = evaluator_details

    names = [d["name"] for d in evaluator_details if d.get("name")]
    async for ev in emitter.result(
        f"List contains {len(evaluator_details)} evaluators: {', '.join(names) if names else '(empty)'}"
        + (f" — {err}" if err else "")
    ):
        yield ev

    async for ev in emitter.phase_end():
        yield ev


async def _phase_4(emitter: SnippetEmitter, client: Any, project_id: str, state: dict) -> AsyncIterator[str]:
    async for ev in emitter.phase_start("phase-4-eval-execute", 4, 4):
        yield ev

    async for ev in emitter.code(
        "from browserstack_ai_sdk import Evaluate\n"
        "\n"
        "for row in rows:\n"
        "    Evaluate.evaluation_execution.evaluate({\n"
        '        "evaluators": evaluators_for_execution,\n'
        '        "data": {\n'
        '            "input": row["input"],\n'
        '            "output": row.get("output", ""),\n'
        '            "expectedOutput": row.get("expected_tool", ""),\n'
        "        },\n"
        '        "concurrency": 1,\n'
        "    })",
        log="Run the evaluator list against each row via evaluation_execution.evaluate",
    ):
        yield ev

    rows = state.get("rows", [])
    evaluators_for_execution = state.get("evaluators_for_execution", [])

    score_total: dict[str, list[float]] = {}
    errors = 0
    for row in rows:
        if not evaluators_for_execution:
            break
        request = {
            "evaluators": evaluators_for_execution,
            "data": {
                "input": row.get("input", ""),
                "output": row.get("output", ""),
                "expectedOutput": row.get("expected_tool", "") or "",
            },
            "concurrency": 1,
        }
        res, perr = _safe(client.evaluate.evaluation_execution.evaluate, request)
        if perr is not None:
            errors += 1
            continue
        results = (res.get("results") or []) if isinstance(res, dict) else (getattr(res, "results", []) or [])
        for r in results:
            metric = (
                (r.get("metricName") or r.get("evaluator")) if isinstance(r, dict)
                else (getattr(r, "metricName", None) or getattr(r, "evaluator", None))
            )
            score = r.get("score") if isinstance(r, dict) else getattr(r, "score", None)
            if metric is not None and isinstance(score, int | float):
                score_total.setdefault(metric, []).append(float(score))

    summary_lines = []
    for metric, vals in score_total.items():
        if vals:
            summary_lines.append(f"{metric}: avg={sum(vals)/len(vals):.2f} ({len(vals)} runs)")
    summary = "; ".join(summary_lines) if summary_lines else "no scores returned"

    async for ev in emitter.result(
        f"Evaluated {len(rows)} rows; errors={errors}; {summary}"
    ):
        yield ev

    async for ev in emitter.phase_end():
        yield ev


async def run_eval_execution(
    emitter: SnippetEmitter, client: Any, project_id: str
) -> AsyncIterator[str]:
    state: dict = {}
    async for ev in _phase_1(emitter, client, project_id, state):
        yield ev
    async for ev in _phase_2(emitter, client, project_id, state):
        yield ev
    async for ev in _phase_3(emitter, client, project_id, state):
        yield ev
    async for ev in _phase_4(emitter, client, project_id, state):
        yield ev
    async for ev in emitter.done():
        yield ev

File: node-backend/src/workflows/evalExecution.ts
/**
 * Workflow 3 — Eval Execution (Confluence v0.1 Script 3).
 *
 * Four phases: source data ingest -> evaluator-list lookup/bootstrap ->
 * evaluator-list inspect -> per-row evaluation execution.
 */
import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import { parse as parseCsv } from "csv-parse/sync";

import type { AISDK } from "@browserstack/ai-sdk";
import { SnippetEmitter } from "../services/snippetEmitter.js";
import { viewInSandbox } from "../services/sdkClient.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const DATA_DIR = path.resolve(__dirname, "..", "..", "data");

const EVAL_LIST_NAME = "support-quality";

const EVALUATOR_SPECS: Array<Record<string, unknown>> = [
  { name: "action-advancement", type: "llm_custom",
    prompt: "Did the assistant advance the user toward their goal? 1 if yes, 0 if no." },
  { name: "tool-selection-quality", type: "llm_custom",
    prompt: "Did the assistant pick the right tool? 1 if yes, 0 if no." },
  { name: "helpfulness", type: "llm_custom",
    prompt: "Rate helpfulness from 1 (unhelpful) to 5 (very helpful)." },
  { name: "latency-under-2s", type: "code", language: "javascript",
    code: "function evaluate(trace) { return (trace.latency_ms ?? 9999) < 2000 ? 1 : 0; }" },
];

async function safe<T>(fn: () => Promise<T> | T): Promise<[T | null, string | null]> {
  try { return [await fn(), null]; }
  catch (err) { return [null, err instanceof Error ? err.message : String(err)]; }
}
function getId(x: unknown): string {
  if (x && typeof x === "object") {
    const r = x as Record<string, unknown>;
    if (typeof r.id === "string") return r.id;
  }
  return "n/a";
}
function getName(x: unknown): string | undefined {
  if (x && typeof x === "object") {
    const r = x as Record<string, unknown>;
    if (typeof r.name === "string") return r.name;
  }
  return undefined;
}

interface Row { input: string; expected_tool?: string | null; output?: string; scope?: string }

export async function runEvalExecution(
  emitter: SnippetEmitter,
  client: AISDK,
  projectId: string,
): Promise<void> {
  interface EvaluatorEntry {
    metricName?: string;
    displayName?: string;
    family?: string;
    runtimeProvider?: string;
    evaluatorId: string;
    params: Record<string, unknown>;
    codeText?: string;
    codeLanguage?: string;
  }
  const state: {
    rows: Row[];
    evaluatorListId: string;
    evaluatorsForExecution: EvaluatorEntry[];
    evaluatorDetails: Array<{ id: string; name?: string }>;
  } = {
    rows: [],
    evaluatorListId: "n/a",
    evaluatorsForExecution: [],
    evaluatorDetails: [],
  };

  // ---- Phase 1: Source data ingest ----
  emitter.phaseStart("phase-1-source-data", 1, 4);
  emitter.code(
    `const text = fs.readFileSync("data/dataset.csv", "utf8");\n` +
    `const rows = parse(text, { columns: true });\n` +
    `console.log(rows.slice(0, 3));`,
    { log: "Load eval input rows from CSV" },
  );
  const csvPath = path.join(DATA_DIR, "dataset.csv");
  if (fs.existsSync(csvPath)) {
    const text = fs.readFileSync(csvPath, "utf8");
    const rows = parseCsv(text, { columns: true, skip_empty_lines: true }) as Record<string, string>[];
    state.rows = rows.map((r) => ({
      input: r.input ?? "",
      expected_tool: r.expected_tool ?? "",
      scope: r.scope ?? "in",
    }));
  } else {
    const jsonPath = path.join(DATA_DIR, "dataset.json");
    if (fs.existsSync(jsonPath)) {
      const data = JSON.parse(fs.readFileSync(jsonPath, "utf8")) as { in_scope?: Array<{ query: string; expected_tool: string | null }> };
      state.rows = (data.in_scope ?? []).map((r) => ({
        input: r.query, expected_tool: r.expected_tool ?? "", scope: "in",
      }));
    }
  }
  emitter.result({
    log: `Loaded ${state.rows.length} rows; first 3: ${JSON.stringify(state.rows.slice(0, 3))}`,
  });
  emitter.phaseEnd();

  // ---- Phase 2: Evaluator list lookup or bootstrap ----
  emitter.phaseStart("phase-2-evaluator-list", 2, 4);
  emitter.code(
    `import { Evaluate } from "@browserstack/ai-sdk";\n` +
    `\n` +
    `const existingLists = await Evaluate.evaluatorList.list();`,
    { log: "Look up 'support-quality'" },
  );
  const elApi = (client as unknown as { evaluate: { evaluatorList: {
    list: () => Promise<{ evaluators?: unknown[] }>;
    create: (body: Record<string, unknown>) => Promise<{ id?: string }>;
    get: (id: string) => Promise<unknown>;
  } } }).evaluate.evaluatorList;
  const evApi = (client as unknown as { evaluate: { evaluator: {
    list: (opts?: Record<string, unknown>) => Promise<{ evaluators?: unknown[]; data?: unknown[] }>;
    create: (body: Record<string, unknown>) => Promise<{ id?: string }>;
  } } }).evaluate.evaluator;

  const [listed] = await safe(() => elApi.list());
  // Real shape: { evaluators: [...], totalCount: number } — handle dict and
  // array forms. Reading only `.evaluators` on a non-conforming response left
  // us with [] every time, so the bootstrap path kept running and 409'ing.
  const lists: unknown[] = Array.isArray(listed)
    ? (listed as unknown[])
    : (listed?.evaluators ?? (listed as { data?: unknown[] })?.data ?? []);
  let found = lists.find((l) => getName(l) === EVAL_LIST_NAME);

  if (!found) {
    emitter.code(
      `import { Evaluate } from "@browserstack/ai-sdk";\n` +
      `\n` +
      `// 'support-quality' missing — bootstrap inline so this workflow stays\n` +
      `// self-contained when Workflow 1 hasn't been run. Reuse evaluators if\n` +
      `// they already exist (server rejects duplicate names with 409).\n` +
      `const existing = await Evaluate.evaluator.list({ limit: 100 });\n` +
      `const byName = new Map((existing.evaluators ?? []).map(e => [e.name, e.id]));\n` +
      `const evaluatorIds = [];\n` +
      `for (const spec of EVALUATOR_SPECS) {\n` +
      `  if (byName.has(spec.name)) { evaluatorIds.push(byName.get(spec.name)); continue; }\n` +
      `  const e = await Evaluate.evaluator.create(spec);\n` +
      `  evaluatorIds.push(e.id);\n` +
      `}\n` +
      `await Evaluate.evaluatorList.create({\n` +
      `  name: "${EVAL_LIST_NAME}",\n` +
      `  evaluators: evaluatorIds.map(id => ({ evaluatorId: id, params: [] })),\n` +
      `});`,
      { log: "Bootstrap evaluators + list (reuse-if-exists)" },
    );

    // Reuse-if-exists: list current evaluators by name first, then only
    // create the ones that are missing. Server rejects duplicate names with
    // 409 "already exists", which would leave us with an empty list and a
    // broken eval-list create below.
    const [existingListed] = await safe(() => evApi.list({ limit: 100 }));
    const existingEvals: unknown[] = Array.isArray(existingListed)
      ? (existingListed as unknown[])
      : (existingListed?.evaluators ?? existingListed?.data ?? []);
    const existingByName = new Map<string, string>();
    for (const e of existingEvals) {
      const n = getName(e);
      const id = getId(e);
      if (n && id !== "n/a") existingByName.set(n, id);
    }

    const evalIds: string[] = [];
    for (const spec of EVALUATOR_SPECS) {
      const specName = spec.name as string;
      if (existingByName.has(specName)) {
        evalIds.push(existingByName.get(specName) as string);
        continue;
      }
      const [res, cerr] = await safe(() => evApi.create(spec));
      const id = getId(res);
      if (id !== "n/a") {
        evalIds.push(id);
      } else {
        emitter.code(
          `// evaluator.create(${specName}) failed: ${cerr}`,
          { log: `evaluator.create(${specName}) failed: ${cerr}` },
        );
      }
    }
    const [created, cerr] = await safe(() => elApi.create({
      name: EVAL_LIST_NAME, evaluators: evalIds.map((id) => ({ evaluatorId: id, params: [] })),
    }));
    state.evaluatorListId = getId(created);
    emitter.result({
      log: cerr ? `Create failed: ${cerr}` : `Created (id=${state.evaluatorListId})`,
      viewInSandbox: { label: "View evaluator list", url: viewInSandbox(projectId, "evaluator-list", state.evaluatorListId) },
    });
  } else {
    state.evaluatorListId = getId(found);
    emitter.code(`// Found existing list (id=${state.evaluatorListId}); reusing`, { log: "Reusing list" });
    emitter.result({
      log: `Reused (id=${state.evaluatorListId})`,
      viewInSandbox: { label: "View evaluator list", url: viewInSandbox(projectId, "evaluator-list", state.evaluatorListId) },
    });
  }
  emitter.phaseEnd();

  // ---- Phase 3: Evaluator list inspect ----
  emitter.phaseStart("phase-3-evaluator-list-get", 3, 4);
  emitter.code(
    `import { Evaluate } from "@browserstack/ai-sdk";\n` +
    `\n` +
    `const fetched = await Evaluate.evaluatorList.get("${state.evaluatorListId}");\n` +
    `// Walk evaluatorConfigs[] to get each evaluatorId, then fetch full evaluator\n` +
    `// rows so we can build the rich payload for Phase 4.\n` +
    `const configs = fetched?.evaluatorConfigs ?? [];\n` +
    `const evaluatorIds = configs.map(c => c?.evaluatorId).filter(Boolean);\n` +
    `const evaluatorsForExecution = [];\n` +
    `for (const eid of evaluatorIds) {\n` +
    `  const ev = await Evaluate.evaluator.get(eid);\n` +
    `  const entry = {\n` +
    `    metricName: ev?.name,\n` +
    `    displayName: ev?.name,\n` +
    `    family: ev?.family,\n` +
    `    runtimeProvider: ev?.runtimeProvider,\n` +
    `    evaluatorId: eid,\n` +
    `    params: {},\n` +
    `  };\n` +
    `  if (ev?.code) entry.codeText = ev.code;\n` +
    `  if (ev?.language) entry.codeLanguage = ev.language;\n` +
    `  evaluatorsForExecution.push(entry);\n` +
    `}`,
    { log: "Walk evaluatorConfigs, fetch each evaluator, build payload for evaluationExecution" },
  );

  const evGetApi = (client as unknown as { evaluate: { evaluator: {
    get: (id: string) => Promise<unknown>;
  } } }).evaluate.evaluator;

  const [fetched, ferr] = await safe(() => elApi.get(state.evaluatorListId));
  const configs: unknown[] =
    fetched && typeof fetched === "object"
      ? ((fetched as { evaluatorConfigs?: unknown[] }).evaluatorConfigs ?? [])
      : [];
  const evaluatorIds: string[] = configs
    .map((c) =>
      c && typeof c === "object"
        ? (c as { evaluatorId?: string }).evaluatorId
        : undefined,
    )
    .filter((x): x is string => typeof x === "string");

  for (const eid of evaluatorIds) {
    const [evObj] = await safe(() => evGetApi.get(eid));
    if (!evObj || typeof evObj !== "object") continue;
    const r = evObj as Record<string, unknown>;
    const name = typeof r.name === "string" ? r.name : undefined;
    const family = typeof r.family === "string" ? r.family : undefined;
    const runtime = typeof r.runtimeProvider === "string" ? r.runtimeProvider : undefined;
    const codeText = typeof r.code === "string" ? r.code : undefined;
    const codeLang = typeof r.language === "string" ? r.language : undefined;
    state.evaluatorDetails.push({ id: eid, name });
    const entry: EvaluatorEntry = {
      metricName: name,
      displayName: name,
      family,
      runtimeProvider: runtime,
      evaluatorId: eid,
      params: {},
    };
    if (codeText) entry.codeText = codeText;
    if (codeLang) entry.codeLanguage = codeLang;
    state.evaluatorsForExecution.push(entry);
  }

  const names = state.evaluatorDetails
    .map((d) => d.name)
    .filter((n): n is string => typeof n === "string");
  emitter.result({
    log: `List contains ${state.evaluatorDetails.length} evaluators: ${names.join(", ") || "(empty)"}${ferr ? ` — ${ferr}` : ""}`,
  });
  emitter.phaseEnd();

  // ---- Phase 4: Per-row evaluation ----
  emitter.phaseStart("phase-4-eval-execute", 4, 4);
  emitter.code(
    `import { Evaluate } from "@browserstack/ai-sdk";\n` +
    `\n` +
    `for (const row of rows) {\n` +
    `  await Evaluate.evaluationExecution.evaluate({\n` +
    `    evaluators: evaluatorsForExecution,\n` +
    `    data: {\n` +
    `      input: row.input,\n` +
    `      output: row.output ?? "",\n` +
    `      expectedOutput: row.expected_tool ?? "",\n` +
    `    },\n` +
    `    concurrency: 1,\n` +
    `  });\n` +
    `}`,
    { log: "Run the evaluator list against each row via evaluationExecution.evaluate" },
  );

  const evExecApi = (client as unknown as { evaluate: { evaluationExecution: {
    evaluate: (req: Record<string, unknown>) => Promise<{ results?: Array<{ metricName?: string; evaluator?: string; score?: number }> }>;
  } } }).evaluate.evaluationExecution;

  const totals: Record<string, number[]> = {};
  let errors = 0;
  for (const row of state.rows) {
    if (state.evaluatorsForExecution.length === 0) break;
    const [res, perr] = await safe(() => evExecApi.evaluate({
      evaluators: state.evaluatorsForExecution,
      data: {
        input: row.input,
        output: row.output ?? "",
        expectedOutput: row.expected_tool ?? "",
      },
      concurrency: 1,
    }));
    if (perr) { errors += 1; continue; }
    const results = res?.results ?? [];
    for (const r of results) {
      const metric = r.metricName ?? r.evaluator;
      if (metric && typeof r.score === "number") {
        (totals[metric] ??= []).push(r.score);
      }
    }
  }
  const summaryLines = Object.entries(totals)
    .filter(([, v]) => v.length)
    .map(([k, v]) => `${k}: avg=${(v.reduce((a, b) => a + b, 0) / v.length).toFixed(2)} (${v.length} runs)`);
  const summary = summaryLines.length ? summaryLines.join("; ") : "no scores returned";
  emitter.result({ log: `Evaluated ${state.rows.length} rows; errors=${errors}; ${summary}` });
  emitter.phaseEnd();
}
